hbase-client/src/main/java/org/apache/hadoop/hbase/client/CatalogReplicaLoadBalanceReplicaSimpleSelector.java

```diff
@@ -0,0 +1,273 @@
+    byte[] startKey = loc.getRegion().getStartKey();
+    tableCache.cache.putIfAbsent(startKey,
+      new StaleLocationCacheEntry(fromReplicaId, loc.getRegion().getEndKey()));
+    LOG.debug("Add entry to stale cache for table {} with startKey {}, {}",
+      loc.getRegion().getTable(), startKey, loc.getRegion().getEndKey());
+  }
+
+  private int getRandomNonPrimaryReplicaId() {
+    if (numOfReplicas < 0) {
+      try {
+        Admin admin = conn.toConnection().getAdmin();
+        TableDescriptor tableDescriptor = admin.getDescriptor(tableName);
+        this.numOfReplicas = tableDescriptor.getRegionReplication();
+      } catch (IOException ioe) {
+        LOG.error("Failed to get table {}'s region replication, exception ", tableName, ioe);
+        this.numOfReplicas = 1;
+      }
+    }
+    // In case of no replica configured, return the primary region id.
+    if (this.numOfReplicas <= 1) {
+      return RegionInfo.DEFAULT_REPLICA_ID;
+    }
+    return 1 + ThreadLocalRandom.current().nextInt(this.numOfReplicas - 1);
+  }
+
+  /**
+   * When it looks up a location, it will call this method to find a replica replica region to go.
+   * For a normal case, > 99% of region locations from catalog/meta replica will be up to date.
+   * In extreme cases such as region server crashes, it will depends on how fast replication
+   * catches up.
+   *
+   * @param tablename  table name it looks up
+   * @param row   key it looks up.
+   * @param locateType locateType, Only BEFORE and CURRENT will be passed in.
+   * @return catalog replica id
+   */
+  public int selectReplica(final TableName tablename, final byte[] row,
+    final RegionLocateType locateType) {
+    StaleTableCache tableCache = staleCache.get(tablename);
+
+    // If there is no entry in StaleCache, select a random replica id.
+    if (tableCache == null) {
+      return getRandomNonPrimaryReplicaId();
+    }
+
+    Map.Entry<byte[], StaleLocationCacheEntry> entry;
+    boolean isEmptyStopRow = isEmptyStopRow(row);
+    // Only BEFORE and CURRENT are passed in.
+    if (locateType == RegionLocateType.BEFORE) {
+      entry = isEmptyStopRow ? tableCache.cache.lastEntry() : tableCache.cache.lowerEntry(row);
+    } else {
+      entry = tableCache.cache.floorEntry(row);
+    }
+
+    // It is not in the stale cache, return a random replica id.
+    if (entry == null) {
+      return getRandomNonPrimaryReplicaId();
+    }
+
+    // Check if the entry times out.
+    if ((System.currentTimeMillis() - entry.getValue().getTimestamp()) >=
+      STALE_CACHE_TIMEOUT_IN_MILLISECONDS) {
+      LOG.debug("Entry for table {} with startKey {}, {} times out", tablename, entry.getKey(),
+        entry);
+      tableCache.cache.remove(entry.getKey());
+      return getRandomNonPrimaryReplicaId();
+    }
+
+    byte[] endKey =  entry.getValue().getEndKey();
+
+    // The following logic is borrowed from AsyncNonMetaRegionLocator.
+    if (isEmptyStopRow(endKey)) {
```

Comment:
I think the below checks should be moved before the 'Check if the entry times out' check? Before the below checks, you do not know whether it is the location you are looking for...

Reply:
The entry found is a possible match before these checks. The 'timeout' check is to make sure the possible match does not timeout. If it times out, it is not necessary to perform these checks. Logically, I think both orders are correct. In case of a 'timeout' entry, it does not need to perform these checks. what do you think?

Reply:
Please add more comments to explain the logic here. I think you should at least say this:

"Check the timeout first as comparing long is faster than comparing byte arrays(in most cases), and it could remove the stale entries more quickly. It is not a problem if this is not the entry for us, as if it times out, we will go to random selection, if it does not time out, we will go to the next check and know it is not for us, then we will also go to random selection."

Reply:
Sure, will do.